### PR TITLE
Stats and WSDA Movement

### PR DESCRIPTION
**Pull request by Benayash**
Added basic stats and WSDA movement
Stats are basicly used as a template to add stats to whatever we want - NPC, Enemy and players.
Added everything in a sample scene.
Also added scripts that are not in use but will be used later like mouse and game manager.
Scene has a sphere as player and "player stats" as his stats.

### CODE REVIEW

File: Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats_SO.cs [hunk 1]
```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[CreateAssetMenu(fileName = "New Stats", menuName ="NPC/Stats/Character Stats")]
+public class CharacterStats_SO : ScriptableObject
```
Comment (maciej-plonka) [inline Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats_SO.cs:7]: Strange naming convention, ScriptableObject is still a class - use PascalCase

File: Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+public class CharacterStats : MonoBehaviour
+{
+
+    #region Defenitions
+
+    public CharacterStats_SO characterDefenition;
+
+
+    public bool isInCombat = false;
```
Comment (maciej-plonka) [inline Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats.cs:15]: Why is CharacterStats aware of whether Character is in combat or not

Comment (Benayash) [reply Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats.cs:15]: It said top reduce Healthregen to 25% when in combat so I prepared this in case we will have a state of "combat", then to add a function for reducing the regen as long as the state is "in combat".

File: Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats_SO.cs (hunk 1 above)
Comment (Benayash) [reply Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats_SO.cs:7]: Well that's how I learned it from unity learn ^^
There is an SO(ScriptableObject) for the creating of the object's stats and regular Charstats to use it the SO as an object.

File: Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats.cs (hunk 2 above)
Comment (maciej-plonka) [reply Assets/Scripts/Aanallein/Scripts/Stats/CharacterStats.cs:15]: we could use either Decorator pattern or different construct here to get that working.
stat = base stat + stat gained from lvlup + stat from buffs + stat from items - stat from debuff... etc.
The structure of your abstraction can be totally different than what user / player will see. They will only be shown the projection of system's current state.

**Comment by maciej-plonka** [comment]
Maybe we could somehow generify stats?
Will all characters have the same set of stats?
Taking saving / loading into consideration what would be the most optimal way to keep stats?

**Comment by Benayash** [comment]
> Please remove unnecessary binary files.
> Using ScriptableObject for player stats is a great idea.
> CharacterStats shouldn't keep track of things like "if player is in combat", "current hp".
> Character's resources such as mana, stamina, health points should only depend on stats.

The current HP and current armor and stuff is essential because it keeps all the data in one place so it will be easy to link it to the game objects and reference it in the scripts.
Current and base are for example the base attackpower to be saved in order to revert to it from equipping / unequipping and stuff like that. like the base is what strength gives, and the current is the base + all buffs.

>Maybe we could somehow generify stats?
I dont think its a good option as we give everyone different stats and it give the option to customize battle NPC and enemy stats with ease while also giving basic stats to whomever we with (through the start function in charstats script)

>Will all characters have the same set of stats?

It can be generified (made generic :P) with the start function, so like random NPC can have fixed stats.
Also if we dont need the stats for a certain NPC we can just make it a 0.
Not all stats will be visble though AI and we can pick exactly which will be shown so I dont think background 0 stats is an issue.
It makes it easier to add stats like after 6 months we want to add "fire resistance" or something, we just add it in 1 file instead of adding it to every npc that needs it manually..

>Taking saving / loading into consideration what would be the most optimal way to keep stats?

It can be used with keeping files that save this data but I have yet to learn how to do it..
Basically you define a file format (xml or the sort) and at awake you set all the values of the stats to the value in the file. Shouldnt be that hard.


P.S how do I change the settings of the GIT to upload only files and not TEMP files?
